# Run QC

## Background

This script was built to take in the `bam` file and to preform quality control analysis on the virus library.

### Purpose

This script is the third step for the vRAPID pipeline for the Bakel Lab. It uses `bamtools` to split the input file, before running `picard`. Then, using `samtools depth`, the coverage information is obtained. Then `kraken2` is used on a subsampled FASTQ files, outputting the `kraken_report.out`. Lastly, it calls on two other scripts, [`taxanomic_breakdown.R`] and [`plot-coverage-report.R`], before outputting the final `qualityControl.pdf` for each sample.

## Script Usage

### Execution

`run_QC.py -rd <path-to-repo-dir> -i <sample_folder> -kdb /sc/arion/projects/PVI/db/minikraken_8GB_20200312 -r <reference-genome-FASTA> -pf <forward-primer-FASTA> -pr <reverse-primer-FASTA> -v <virus-type> -pc <plotcoverage-script-path> -tb <taxanomic-breakdown-script-path> `

#### Snakemake

The script can be found in the `Snakefile` found in the `workflow` directory, under the `QC_analysis` rule. All necessary input, parameters, output, and log files are defined under that rule. Note that the snakemake rule also outputs a log file for each sample. Snakemake's log files can always be found in the `log` directory, and identified per sample in the following format:

`03_<sample-name>.QC.snakemake.log`

#### Directory structure

For the script to be successfully excuted, the following directory structure is required:

```bash
<Run-ID>
├── <sample_ID>
│   ├── <01_fastqs>
│   │   ├── <sample_ID>_1.fastq.gz
│   │	│	├── <sample_ID>_2.fastq.gz
│   ├── <02_assembly>
│   │   ├── <sample_ID>.fasta
│   │	│	├── <sample_ID>_ref.bam
│   │	│	├── <sample_ID>_ref.bam.bai
│   │	│	├── *_pilon.*
│   │	│	├── <reference-FASTA-header>_shovill
│   │	│	├── prokka
│   ├── <04_variants>
│   │   ├── <sample_ID>.<reference-fasta-header>_variable_bases.tsv
│   │	│	├── <reference-fasta-header>.final.fna
│   │	│	├── <reference-fasta-header>_pileup.gz
│   │	│	├── pileup.gz
```

#### *Note*

`taxanomic_breakdown.R`

Parses the `kraken` output before plotting a breakdown of the taxanomic reads of the viruses of interest within the sample.

`plot-coverage.R` 

Parses the TSV generated by running `samtools mpileup` in the variant analysis step, before plotting the coverage across the genome.

### Output

The script will create a new subdirectory with the `kraken`  output files, coverage report information, `<reference-FASTA-header>_coverage.txt`, mapped read information, `<sample_ID>_refbam.flagstat`, and final figures of the QC analysis, `<sample_ID>_quality_control.pdf`. 

 ```bash
<Run-ID>
├── <sample_ID>
│   ├── <01_fastqs>
│   │  ├── <sample_ID>_1.fastq.gz
│   │	│	├── <sample_ID>_2.fastq.gz
│   ├── <02_assembly>
│   │  ├── <sample_ID>.fasta
│   │	│	├── <sample_ID>_ref.bam
│   │	│	├── <sample_ID>_ref.bam.bai
│   │	│	├── *_pilon.*
│   │	│	├── <reference-FASTA-header>_shovill
│   │	│	├── prokka
│   ├── <04_variants>
│   │  ├── <sample_ID>.<reference-fasta-header>_variable_bases.tsv
│   │	│	├── <reference-fasta-header>.final.fna
│   │	│	├── <reference-fasta-header>_pileup.gz
│   │	│	├── pileup.gz
│   ├── <03_qualityControl>
│   │  ├──  <reference-FASTA-header>_coverage.txt
│   │  ├──  <reference-FASTA-header>_report.txt
│   │  ├── <sample_ID>_kraken
│   │  ├── <sample_ID>_kraken_input.fastq.gz
│   │  ├── <sample_ID>_kraken_report.out
│   │  ├── <sample_ID>_quality_control.pdf
│   │  ├── <sample_ID>_refbam.flagstat
 ```

